README.md
# Arduino-Blinkenlight

[![tests](https://github.com/tfeldmann/Arduino-Blinkenlight/actions/workflows/tests.yml/badge.svg)](https://github.com/tfeldmann/Arduino-Blinkenlight/actions/workflows/tests.yml)

Featured on [hackster.io](https://www.hackster.io/news/thomas-feldmann-s-arduino-blinkenlight-brings-you-smooth-non-blocking-status-led-flashing-and-fading-1d533e3cd20c)!

## 🚨 Supercharge your status-LEDs 🚨

> This library gives you non-blocking blinking patterns and smooth fade effects for your
> LEDs, buzzers or any other status indicators

<pre align="center">
<strong>Achtung Alles Lookenskeepers!</strong>
Das computermachine ist nicht fuer gefingerpoken und
mittengrabben. Ist easy schnappen der springenwerk,
blowenfusen und poppencorken mit spitzensparken. Ist nicht
fuer gewerken bei das dumpkopfen. Das rubbernecken
sichtseeren keepen das cotten-pickenen hans in das pockets
muss; <i>relaxen und watchen das blinkenlichten.</i>
</pre>

## Features

- On / off / toggle
- Infinite blinking
- Blink patterns (for example to blink an error code)
- Single flashes and pauses, resuming the previous mode

Other goodies:

- Completely non-blocking (no `delay()`)
- Super-nice fading effects (optional) with logarithmic brightness compensation for LEDs
- Lightweight
- All parameters can be adjusted on the fly without visual hickups
- Easily extendable to control components via SPI / CAN / I2C / UART ...
- Comes with good-looking defaults (at least I tried)
- Tries to be user friendly (for example calling `blink()` while blinking does not
  restart the cycle)
- Supports active low (inverted) logic
- Can be used for buzzers as well

## Example

This example blinks the built-in LED on pin 13 in the following pattern:

![pattern](docs/SpeedSettings.png)

```C
#include <Fadinglight.h>

Fadinglight led(13);

void setup()
{
    led.pattern(2, 3);  // blink 2x, then 3x
}

void loop()
{
    led.update();
}
```

Easy, uh? It's not only blinking, it does so with smooth fading effects and
logarithmic LED brightness compensation. Your boards never looked more professional! /s

> Note: If you don't love the fading effects, just use the `Blinkenlight`-class instead
> of `Fadinglight`.

## Installation

### Platform.io

Blinkenlight is available [in the platform.io registry](https://platformio.org/lib/show/13287/Blinkenlight)

```
pio lib install "tfeldmann/Blinkenlight"
```

### Arduino library manager

`Blinkenlight` is available in the Arduino library manager.

![Blinkenlight in the Arduino library manager](docs/arduino-library.png)

## Full API

### Instantiation

```C
// Without fading effect:
#include <Blinkenlight.h>
Blinkenlight myLed(13);  // Blinkenlight(int pin, bool invert = false);

// With fading effect:
#include <Fadinglight.h>
Fadinglight myLed(13);  // Fadinglight(int pin, bool logarithmic = true, int fade_speed = 30);

// now in your code you can do:
myLed.off();
myLed.blink();
myLed.pattern(3, 5, SPEED_FAST);
// ... and so on (see below)
```

### Basic usage (what you can do with your light)

```C
// set permanently ON
void on();

// set permanently OFF
void off();

// toggle between on / off
void toggle();

// set ON / OFF permanently
void permanent(bool enable);

// blink infinitely
void blink();
void blink(SpeedSetting speed);

// blink `num` times, then long pause
// repeats, if `repeat` is set, OFF otherwise.
void pattern(int num, bool repeat = true);
void pattern(int num, SpeedSetting speed, bool repeat = true);

// blink `num1` times, short pause, blink `num2` times, long pause
// repeats, if `repeat` is set, OFF otherwise.
void pattern(int num1, int num2, bool repeat = true);
void pattern(int num1, int num2, SpeedSetting speed, bool repeat = true);

// turn ON for the given duration in ms. Continues in the previous mode afterwards.
void flash(uint16_t duration_ms);

// turn OFF for the given duration in ms. Continues in the previous mode afterwards.
void pause(uint16_t duration_ms);

// `true` if the Blinkenlight is currently blinking, showing a pattern, flashing or pausing
bool isOn();

// You must call this in your loop!
// Returns the current value of the light (LOW / HIGH).
// - You can ignore that if you want.
int update();
```

### Blinking speed and timing parameters

```C
// setup the timing parameters
void setSpeed(SpeedSetting setting);

// Available by default: SPEED_RAPID, SPEED_FAST, SPEED_SLOW
myLed.setSpeed(SPEED_RAPID);
myLed.setSpeed(SPEED_FAST);
myLed.setSpeed(SPEED_SLOW);

// .. or use your own settings. SpeedSetting is a struct:
typedef struct
{
    uint16_t on_ms;         // ON time in milliseconds
    uint16_t off_ms;        // OFF time in milliseconds
    uint16_t pause_ms;      // pause between `num1` and `num2` in pattern()
    uint16_t ending_ms;     // ending after a full pattern
} SpeedSetting;

// so you can set it up like this:

SpeedSetting mySettings = {
    .on_ms = 100,
    .off_ms = 100,
    .pause_ms = 2000,
    .ending_ms = 2000,
};

// ... alternatively you can setup the speed settings directly with this method
void setSpeed(
    uint16_t on_ms,
    uint16_t off_ms,
    uint16_t pause_ms,
    uint16_t ending_ms);

// ... or by providing a single value, the other values are inferred from that by
// some internal factors
void setSpeed(uint16_t on_ms);

// Hint: You can also modify the values directly - even on the fly - e.g.:
myLed.settings.on_ms = 250;
myLed.settings.pause_ms = 2000;
```

## Fading effect and brightness compensation

LED brightness is not linear to the PWM value you set to the pin.
This library uses a logarithmic dimming curve to create natural looking fading effects.

![logarithmic fading](docs/logarithmic.png)

## I have a status indicator controlled via CAN / I2C / SPI / ... What can I do?

No problem! You have two options.

- Use the generic `BaseBlinker` class from `<BaseBlinker.h>`. The `.update()`-method
  returns a boolean whether the status is currently `HIGH` or `LOW`. You can then send
  this value to your status Blinkenlight (see `examples/GenericBlink`).
  Use the `BaseFader` class if you want fading effects. Here the `update` method
  returns an integer `0..255`.

- Subclass the `BaseBlinker` class with custom logic. This is what `Blinkenlight` does
  internally (see `src/Blinkenlight.h`). Have a look at the `SerialBlink` example!

## Why

I created this library because I like to blink my error codes to the user.
So for example blink two times, then three times for error 23 - you get the idea.

Over the time I found it super useful to handle basically all buzzers, lights and LEDs.
I have multiple applications where it controls status lights via CAN.


CHANGELOG.md
# Changelog

## v2.3.0 (2021-12-20)

- `logarithmic` is now `true` by default
- Add `pattern` and `flash` method with `SpeedSetting` parameter
- Added example `FadeComparison`

## v2.2.0 (2021-12-17)

- Fixes a bug where no default speed is set
- Fixes a bug in the SerialBlink example where `Serial.begin()` is not called
- `BaseFader` now supports custom value ranges (default = 255)
- Small code speedup in `BaseFader`

## v2.1.0 (2021-12-16)

- Removed default speeds from methods

## v2.0.1 (2021-12-16)

- Renamed library to `Blinkenlight`

## v2.0.0 (2021-12-16)

- **Revamped API!**
- New: writes the state only on changes, not on every update
- Renamed the classes and header files:
  - `BlinkenlightPin` -> `Blinkenlight`
  - `FadingBlinkenlightPin` -> `Fadinglight`
  - `Blinkenlight` -> `BaseBlinker`
  - `FadingBlinkenlight` -> `BaseFader`
- Code cleanups
- Uses GitHub Actions for tests

## v1.2.0 (2021-11-15)

- Renamed `configure`-method

## v1.1.0 (2021-11-12)

- renamed `count()` method to `pattern()`

## v1.0.1 (2021-11-12)

- Initial release


src/BaseFader.h
#pragma once
#include <Arduino.h>

#include "BaseBlinker.h"

// logarithmic led brightness curve
// clang-format off
static const uint8_t LED_LOG_CURVE[256] = {
   0,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
   1,  1,  1,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
   2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  3,  3,  3,  3,  3,
   3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  4,  4,  4,  4,  4,  4,
   4,  4,  4,  4,  4,  4,  5,  5,  5,  5,  5,  5,  5,  5,  5,  6,
   6,  6,  6,  6,  6,  6,  6,  7,  7,  7,  7,  7,  7,  8,  8,  8,
   8,  8,  8,  9,  9,  9,  9,  9, 10, 10, 10, 10, 10, 11, 11, 11,
  11, 12, 12, 12, 12, 13, 13, 13, 14, 14, 14, 14, 15, 15, 15, 16,
  16, 16, 17, 17, 18, 18, 18, 19, 19, 20, 20, 21, 21, 21, 22, 22,
  23, 23, 24, 24, 25, 25, 26, 27, 27, 28, 28, 29, 30, 30, 31, 32,
  32, 33, 34, 35, 35, 36, 37, 38, 39, 39, 40, 41, 42, 43, 44, 45,
  46, 47, 48, 49, 50, 51, 52, 53, 55, 56, 57, 58, 59, 61, 62, 63,
  65, 66, 68, 69, 71, 72, 74, 76, 77, 79, 81, 82, 84, 86, 88, 90,
  92, 94, 96, 98,100,102,105,107,109,112,114,117,119,122,124,127,
 130,133,136,139,142,145,148,151,155,158,162,165,169,172,176,180,
 184,188,192,196,201,205,210,214,219,224,229,234,239,244,250,255,
};
// clang-format on

class BaseFader : public BaseBlinker
{
public:
    BaseFader(bool logarithmic = true, int fadeSpeed = 30, int range = 255)
        : BaseBlinker(),
          output_(0),
          lastUpdate_(millis()),
          logarithmic_(logarithmic),
          fadeSpeed_(abs(fadeSpeed)),
          range_(range)
    {
    }

    int update()
    {
        int state = BaseBlinker::update();
        uint32_t ms = millis();

        if (ms - lastUpdate_ > 10)
        {
            lastUpdate_ = ms;
            int diff = state * range_ - output_;

            // we only call `write` on changes
            if (diff)
            {
                output_ += constrain(diff, -fadeSpeed_, fadeSpeed_);
                int result = logarithmic_ ? LED_LOG_CURVE[output_] : output_;
                write(result);
                return result;
            }
        }
        return logarithmic_ ? LED_LOG_CURVE[output_] : output_;
    }

    virtual void write(int state)
    {
        // write() is only called on changes.
        // You can override it and set your Blinkenlight here.
        (void)state;
    }

    void set(int state) override
    {
        // we have to override this method here to prohibit the `BaseBlinker` from
        // calling `write`. Otherwise we write unwanted 0 and 1 instead of fading.
        state_ = state;
    }

private:
    int output_;
    uint32_t lastUpdate_;

    bool logarithmic_;
    int fadeSpeed_;
    int range_;
};
